scripts/abi_artifact_generator/end_to_end.md

```diff
@@ -0,0 +1,308 @@
+
+These two files together (with the help of `functions.c` file that provides some variable definitions, basic endianness verification, etc) are compiled into the `"functions binary"` (see the graph [here](./templates/) for how it fits into the big picture).
+The debug information from the binary is then extracted to be used as the base of the test - since it is what's used to produce the models (look into the `ImportBinary` analysis for details) used until the very end.
+
+On top of the binary provided to the `ImportBinary` analysis, there are two more distinct ones that are used in a tandem: the `"mmap'ed binary"` and the `"runner binary"`. Let's look into them one at a time (see the same graph [here](./templates/) for the big picture).
+
+`"mmap'ed binary"` boils down to two distinct source files: one generated in assembly containing `setup_{{ test_name }}` function definitions for the argument tests, and the other one containing everything else (it's generated in C, and on top of containing bodies of `setup_{{ test_name }}`s for return value tests it also includes `functions.inc`).
+
+`setup_arguments.S` (the first one) is heavily dependent on the architecture, but for the sake of simplicity, let's only look into a single one (`x86_64`) for the current example. Let's look into the example function bit by bit:
+
+It starts by saving the return value address (it's architecture-dependent where it's location, for `x86_64` it's the top of the stack) into a dedicated `saved_return_address` global memory location.
+```asm
+.global setup_args
+setup_args:
+  pop rax
+  mov qword ptr [saved_return_address], rax
+```
+
+Then it pushes randomly generated data from `randomized_state` onto the stack one word at a time, until the expected stack space (defined in the `common.yml`) is filled.
+```asm
+  mov rax, qword ptr [randomized_state + 616]
+  push rax
+  ;; (more pushes omitted)
+  mov rax, qword ptr [randomized_state + 112]
+  push rax
+```
+
+After which it replaces values of all easily accessible the registers with random values from the same `randomized_state` location.
+```asm
+  mov r15, qword ptr [randomized_state + 104]
+  mov r14, qword ptr [randomized_state + 96]
+  mov r13, qword ptr [randomized_state + 88]
+  mov r12, qword ptr [randomized_state + 80]
+  mov r11, qword ptr [randomized_state + 72]
+  mov r10, qword ptr [randomized_state + 64]
+  mov r9, qword ptr [randomized_state + 56]
+  mov r8, qword ptr [randomized_state + 48]
+  mov rdi, qword ptr [randomized_state + 40]
+  mov rsi, qword ptr [randomized_state + 32]
+  mov rdx, qword ptr [randomized_state + 24]
+  mov rcx, qword ptr [randomized_state + 16]
+  mov rbx, qword ptr [randomized_state + 8]
+  mov rax, qword ptr [randomized_state + 0]
+```
+
+Then it contains a call instruction to the test body itself.
+```asm
+  call test_args
+```
+
+In some cases `longjmp` terminates unless the stack is restored beforehand, so do that:
```

Comment:
"terminates" as in "crashes"?

Reply:
Yeah, as far as I remember, it was a signal (probably `segfault`, but don't quote me on that).

Reply:
What I meant is that the term "terminates" does not imply failure. Use "crashes".

Reply:
Ah, right. My bad.